giraf/binaryGiraf/binWriter.go

```diff
@@ -0,0 +1,271 @@
+	return &BinWriter{
+		bg: bgzf.NewWriter(file, 1), //TODO: Play with different levels of concurrency
+	}
+}
+
+// CompressGiraf will encode a giraf file (.giraf) and output a binary giraf file (.giraf.fe)
+func CompressGiraf(filename string) {
+	inputStream := giraf.GoReadToChan(filename)
+	outfile := fileio.EasyCreate(filename + ".fe")
+	defer outfile.Close()
+	writer := NewBinWriter(outfile)
+	var err error
+
+	// Write info from all girafs in inputStream
+	for record := range inputStream {
+		err = WriteGiraf(writer, record)
+		common.ExitIfError(err)
+	}
+
+	// Close writer
+	err = writer.bg.Close()
+	common.ExitIfError(err)
+}
+
+// Byte size of BinGiraf fixed size fields excluding blockSize.
+// Exluded Fields: qName, path, byteCigar, fancySeq.Seq, qual, notes
+var binGirafFixedSize int = 29
+
+// The Write method for the BinWriter struct compresses a single giraf record and writes to file
+func WriteGiraf(bw *BinWriter, g *giraf.Giraf) error {
+	bw.buf.Reset() // clear buffer for new write
+	var currBuf [8]byte
+
+	// encode data for later use
+	fancySeq := getFancySeq(g.Seq, g.Cigar)
+	qual := encodeQual(g.Qual)
+	notes := notesToBytes(g.Notes)
+
+	// blockSize (uint32)
+	size := binGirafFixedSize +
+		len(g.QName) +
+		(4 * len(g.Path.Nodes)) +
+		(3 * len(g.Cigar)) +
+		(8 * len(fancySeq.Seq)) +
+		(3 * len(qual)) +
+		len(notes)
+	binary.LittleEndian.PutUint32(currBuf[:4], uint32(size))
+	bw.buf.Write(currBuf[:4])
+
+	// qNameLen (uint8)
+	bw.buf.Write([]byte{uint8(len(g.QName))})
+
+	// qName (string)
+	bw.buf.Write([]byte(g.QName))
+
+	// flag (uint8)
+	bw.buf.Write([]byte{g.Flag})
+
+	// tStart (uint32)
+	binary.LittleEndian.PutUint32(currBuf[:4], uint32(g.Path.TStart))
+	bw.buf.Write(currBuf[:4])
+
+	// tEnd (uint32)
+	binary.LittleEndian.PutUint32(currBuf[:4], uint32(g.Path.TEnd))
+	bw.buf.Write(currBuf[:4])
+
+	// path ([]uint32
+	binary.LittleEndian.PutUint16(currBuf[:2], uint16(len(g.Path.Nodes)))
+	bw.buf.Write(currBuf[:2]) // pathLen (uint16)
+
+	for _, val := range g.Path.Nodes {
+		binary.LittleEndian.PutUint32(currBuf[:4], val)
+		bw.buf.Write(currBuf[:4]) // path ([]uint32)
+	}
+
+	// cigar ([]cigar.ByteCigar)
+	binary.LittleEndian.PutUint16(currBuf[:2], uint16(len(g.Cigar)))
+	bw.buf.Write(currBuf[:2]) // numCigarOps (uint16)
+
+	for _, val := range g.Cigar {
+		binary.LittleEndian.PutUint16(currBuf[:2], val.RunLen)
+		bw.buf.Write(currBuf[:2])    // byteCigar.RunLen (uint16)
+		bw.buf.Write([]byte{val.Op}) // byteCigar.Op (byte)
+	}
+
+	// fancySeq (dnaThreeBit.ThreeBit)
+	binary.LittleEndian.PutUint32(currBuf[:4], uint32(fancySeq.Len))
+	bw.buf.Write(currBuf[:4]) // fancySeq.Len (uint32)
+
+	for _, val := range fancySeq.Seq {
+		binary.LittleEndian.PutUint64(currBuf[:8], val) // TODO: does not need to be uint64?
+		bw.buf.Write(currBuf[:8])                       // fancySeq.Seq (uint64)
+	}
+
+	// alnScore (int64)
+	binary.LittleEndian.PutUint64(currBuf[:8], uint64(g.AlnScore))
```

Comment:
Do you need the reslicing?  Just curious.

Reply:
Is it faster to make it an array that you re-slice instead of having it be a slice of length 8?